### PR TITLE
Reduce buffer size of unused side of socket

### PR DESCRIPTION
**Pull request by gregcusack**
#### Problem
We allocate large send and receive buffers for all sockets. However, not all sockets read and write to/from the buffers. In fact, 6 sockets are Read Only and 2 are Write Only. Meaning we are allocating memory that is never used. 

#### Summary of Changes
1) Reduce buffer sizes for the unused side of sockets down to 4 MB
2) Update `SocketConfig` to use builder pattern
3) Deprecate unused `net-utils` methods that use default `SocketConfig`

NOTE: i got rid of the default "bind" methods in `net-utils/lib.rs` so that developers are intentional about their socket allocations. 

Services/Sockets
Read/Write
1) Gossip
2) RPC - TCP
3) Ip_echo - TCP
4) Tvu
5) tvu_quic
6) Repair
7) Repair_quic
8) Serve_repair
9) Serve_repair_quic
10) Ancestor_hashes_requests_quic
11) Ancestor_hashes_requests

Read Only
1) Tpu
2) Tpu_forwards
3) Tpu_vote
4) Tpu_quic
5) Tpu_forwards_quic
6) Tpu_vote_quic

Write Only
1) Retransmitter
2) Broadcast

Follow Up PR:
1) Make better sizing decisions for the sockets that don't need a full 128MB buffer

### CODE REVIEW

File: net-utils/src/lib.rs [hunk 1]
```diff
@@ -441,9 +581,15 @@ fn udp_socket_with_config(config: SocketConfig) -> io::Result<Socket> {
 pub fn bind_common_in_range(
     ip_addr: IpAddr,
     range: PortRange,
+    config: Option<SocketConfig>,
```
Comment (gregcusack) [inline net-utils/src/lib.rs:584]: there are a few of these functions i changed to take in an option. Not sure that is the best call. We could:
1) leave it as is, taking in an Option
2) create a new function `bind_common_in_range_with_config` and pass in SocketConfig
3) modify the original function to just take a SocketConfig

File: net-utils/src/lib.rs [hunk 2]
```diff
@@ -481,8 +627,11 @@ pub fn bind_in_range_with_config(
     ))
 }
 
-pub fn bind_with_any_port(ip_addr: IpAddr) -> io::Result<UdpSocket> {
-    let sock = udp_socket(false)?;
+pub fn bind_with_any_port(ip_addr: IpAddr, config: Option<SocketConfig>) -> io::Result<UdpSocket> {
```
Comment (gregcusack) [inline net-utils/src/lib.rs:630]: same question as above with the `Option`

File: net-utils/src/lib.rs [hunk 3]
```diff
@@ -493,10 +642,11 @@ pub fn bind_with_any_port(ip_addr: IpAddr) -> io::Result<UdpSocket> {
     }
 }
 
-// binds many sockets to the same port in a range
+// binds many sockets to the same port in a range with config
 pub fn multi_bind_in_range(
     ip_addr: IpAddr,
     range: PortRange,
+    config: Option<SocketConfig>,
```
Comment (gregcusack) [inline net-utils/src/lib.rs:649]: same question as above with the `Option`

File: net-utils/src/lib.rs [hunk 4]
```diff
@@ -392,12 +401,126 @@ pub fn is_host_port(string: String) -> Result<(), String> {
 #[derive(Clone, Debug)]
 pub struct SocketConfig {
     pub reuseport: bool,
+    #[cfg(not(any(windows, target_os = "ios")))]
+    pub usage: SocketUsage,
+    #[cfg(not(any(windows, target_os = "ios")))]
+    pub recv_buffer_size: usize,
+    #[cfg(not(any(windows, target_os = "ios")))]
+    pub send_buffer_size: usize,
```
Comment (gregcusack) [inline net-utils/src/lib.rs:409]: these `#[cfg...]`s make this code not look great

File: net-utils/src/lib.rs [hunk 5]
```diff
@@ -35,9 +35,18 @@ pub const VALIDATOR_PORT_RANGE: PortRange = (8000, 10_000);
 pub const MINIMUM_VALIDATOR_PORT_RANGE_WIDTH: u16 = 17; // VALIDATOR_PORT_RANGE must be at least this wide
 
 #[cfg(not(any(windows, target_os = "ios")))]
-const DEFAULT_RECV_BUFFER_SIZE: usize = 64 * 1024 * 1024; // 64 MB - Doubled to 128MB by the kernel
+const DEFAULT_RECV_BUFFER_SIZE: usize = 64 * 1024 * 1024; // 64MB - Doubled to 128MB by the kernel
 #[cfg(not(any(windows, target_os = "ios")))]
-const DEFAULT_SEND_BUFFER_SIZE: usize = 64 * 1024 * 1024; // 64 MB - Doubled to 128MB by the kernel
+const DEFAULT_SEND_BUFFER_SIZE: usize = 64 * 1024 * 1024; // 64MB - Doubled to 128MB by the kernel
+#[cfg(not(any(windows, target_os = "ios")))]
+const DEFAULT_UNUSED_BUFFER_SIZE: usize = 128; // 128B - Depending on kernel distribution, this may be increased to a kernel minimum
```
Comment (gregcusack) [inline net-utils/src/lib.rs:42]: the minimum set here varies by kernel version. 
interestingly, I found on my devbox running linux 5.15 that when I set the socket to `WriteOnly` the read buffer got set to 2kB and when I set the socket to `ReadOnly, the write buffer got set to 4kB.

File: connection-cache/src/connection_cache.rs [hunk 6]
```diff
@@ -571,8 +572,11 @@ mod tests {
         fn default() -> Self {
             Self {
                 udp_socket: Arc::new(
-                    solana_net_utils::bind_with_any_port(IpAddr::V4(Ipv4Addr::UNSPECIFIED))
-                        .expect("Unable to bind to UDP socket"),
+                    solana_net_utils::bind_with_any_port(
+                        IpAddr::V4(Ipv4Addr::UNSPECIFIED),
+                        SocketConfigBuilder::new().build(),
+                    )
+                    .expect("Unable to bind to UDP socket"),
```
Comment (gregcusack) [inline connection-cache/src/connection_cache.rs:579]: importing `solana_net_utils` just for this test. so maybe it makes sense to just have a separate function called: `bind_with_any_port_with_config` and leave `bind_with_any_port` as is

File: net-utils/src/lib.rs [hunk 7]
```diff
@@ -392,21 +413,76 @@ pub fn is_host_port(string: String) -> Result<(), String> {
...
+    #[allow(unused_mut)]
+    pub fn usage(mut self, usage: SocketUsage) -> Self {
+        #[cfg(not(any(windows, target_os = "ios")))]
+        {
+            self.usage = usage;
+        }
+        #[cfg(any(windows, target_os = "ios"))]
+        {
+            let _ = usage;
+        }
```
Comment (steviez) [inline net-utils/src/lib.rs:457]: Is this block actually necessary ? Same as similar blocks in other functions

**Review by steviez — COMMENTED** [review]
I haven't gone through fully yet, but I think we could split this PR into two.
- One PR that adds the scaffolding (ie the SocketConfig builder); there shouldn't be any changes in behavior with this initial PR
- One PR that uses the previously landed scaffolding to adjust buffer sizes

The main benefit of doing so would be isolating the changes of behavior (buffer size adjustment) to a smaller diff. This will make it easier to review the more critical piece. Does this seem like a reasonable split ?

File: net-utils/src/lib.rs (hunk 7 above)
Comment (steviez) [reply net-utils/src/lib.rs:457]: I guess maybe it was added to suppress `unused_arg`, but if we're allowing `unused_mut` then we could seemingly do the same thing

Comment (gregcusack) [reply net-utils/src/lib.rs:457]: ya exactly. but ya good point will update

**Comment by gregcusack** [comment]
> I haven't gone through fully yet, but I think we could split this PR into two.
> 
> * One PR that adds the scaffolding (ie the SocketConfig builder); there shouldn't be any changes in behavior with this initial PR
> * One PR that uses the previously landed scaffolding to adjust buffer sizes
> 
> The main benefit of doing so would be isolating the changes of behavior (buffer size adjustment) to a smaller diff. This will make it easier to review the more critical piece. Does this seem like a reasonable split ?

ahhh great idea. yes i will break up into 2 PRs

**Pull request closed without merging**